File: rust/shared/src/read_input.rs
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader, Lines};
use std::path::{Path, PathBuf};

fn get_input_path() -> PathBuf {
    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    let workspace_root = manifest_dir
        .parent()
        .and_then(|p| p.parent())
        .unwrap_or_else(|| Path::new("."))
        .to_owned();

    workspace_root.join("inputs")
}

pub fn read_input(
    year: &str,
    day: &str,
    example: bool,
) -> Result<Lines<BufReader<File>>, Box<dyn Error>> {
    let mut suffix: &str = "";
    if example {
        suffix = "_ex";
    }

    let file_path = get_input_path().join(format!("{}/day{}{}.txt", year, day, suffix));
    let file = File::open(&file_path)?;
    let reader = BufReader::new(file);
    Ok(reader.lines())
}

File: c++/infra/read_input.cpp
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <string>

using namespace std;
using namespace std::filesystem;

auto get_input_path() -> path {
    path current = current_path();
    while (true) {
        if (exists(current / "inputs") && is_directory(current / "inputs"))
            return current / "inputs";

        if (current.has_parent_path())
            current = current.parent_path();
        else
            break;
    }
    throw runtime_error("Inputs not found.");
}

auto read_input(const string& year, const string& day, const bool example) -> ifstream {
    const std::string suffix = example ? "_ex" : "";
    const path filePath = get_input_path() / year / ("day" + day + suffix + ".txt");

    ifstream file(filePath);
    if (!file)
        throw runtime_error("Failed to open file: " + filePath.string());
    return file;
}
